### PR TITLE
[mlir][EmitC] Fix evaluation order of expressions

### PR DESCRIPTION
**Pull request by simon-camp**
Expressions with the same precedence were not parenthesized and therefore were possibly evaluated in the wrong order depending on the shape of the expression tree.

### CODE REVIEW

**Comment by aniragil** [comment]
Thanks for fixing this @simon-camp! 
IINM we can avoid parentheses if not required by associativity, i.e. the expression `(a * b) / d` can be emitted as `a * b / d` due to the operator's left associativity, but best to first fix the bug and later relax if desired.

**Review by aniragil — CHANGES_REQUESTED** [review]

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp [hunk 1]
```diff
@@ -1316,7 +1316,11 @@ LogicalResult CppEmitter::emitOperand(Value value) {
     FailureOr<int> precedence = getOperatorPrecedence(def);
     if (failed(precedence))
       return failure();
-    bool encloseInParenthesis = precedence.value() < getExpressionPrecedence();
+
+    // Sub-expressions with equal or lower precedence need to be paranthesized,
```
Comment (cferry-AMD) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:1320]:
```suggestion
    // Sub-expressions with equal or lower precedence need to be parenthesized,
```

**Review by cferry-AMD — COMMENTED** [review]
Thanks for noticing there were actually two issues -- and for taking care of them! Do you want to add the fix you thought of for the other issue in this PR, or do you want to make another PR for it?

**Comment by simon-camp** [comment]
I'll send another PR for the original issue as that may be need for some discussion.

**Review by mgehre-amd — APPROVED** [review]

**Pull request merged**